File: 2. Hash Maps and Sets/36. Valid Sudoku/Solution.py
from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_sets = [set() for _ in range(9)]
        col_sets = [set() for _ in range(9)]
        box_sets = [[set() for _ in range(3)] for _ in range(3)]

        for row in range(9):
            for col in range(9):
                val = board[row][col]

                if val == '.':
                    continue
                
                if val in row_sets[row]:
                    return False
                
                if val in col_sets[col]:
                    return False
                
                if val in box_sets[row // 3][col // 3]:
                    return False

                row_sets[row].add(val)
                col_sets[col].add(val)
                box_sets[row // 3][col // 3].add(val)
        
        return True

File: 2. Hash Maps and Sets/36. Valid Sudoku/Solution.cpp
#include <vector>
#include <unordered_set>

class Solution {
public:
    bool isValidSudoku(std::vector<std::vector<char>>& board) {
        std::vector<std::unordered_set<char>> row_sets(9);
        std::vector<std::unordered_set<char>> col_sets(9);
        std::vector<std::vector<std::unordered_set<char>>> box_sets(3, std::vector<std::unordered_set<char>>(3));

        for(int row = 0; row < 9; ++row)
        {
            for(int col = 0; col < 9; ++col)
            {
                char val = board[row][col];

                if (val == '.') continue;
                if (row_sets[row].count(val)) return false;
                if (col_sets[col].count(val)) return false;
                if (box_sets[row / 3][col / 3].count(val)) return false;

                row_sets[row].insert(val);
                col_sets[col].insert(val);
                box_sets[row / 3][col / 3].insert(val);
            }
        }
        return true;
    }
};